Update from auto-update/renku-graph-2.27.0 (#2914)
SwissDataScienceCenter/renku: CHANGELOG.rst

.. _changelog:

0.23.0
------

Renku ``0.23.0`` brings number of bug fixes, KG APIs performance improvements as well as support for CLI 1.11.1.

**🐞 Bug Fixes**

* **Knowledge Graph**: Performance improvements for Cross-Entities Search and Project Details APIs.

> * **Knowledge Graph**: Partial support for Workflow File metadata.

Reviewer: Could you be more specific about what this means? What is and is not supported? 
If it would be too long to get into the details here in the Changelog, is there a PR that you could link to from here where the specifics are described?

Reviewer 2: Unfortunately, it's not a simple one. Generally speaking, for the time being, we'll treat the `WorkflowFile` specific `Plan` and `CompositePlan` entities as their non-WorkflowFile versions. As they are more or less the same (the WorklfowFile ones contain some additional properties) there's not much of a problem on the KG side to see them that way. That gives us that all the most essential information related to the WorkflowFile feature will be present in the Triples Store and queryable by our APIs. Obviously, information such as the path to the workflow file yaml or a group of activities generated during a single workflow execution won't be there. This should not be a problem as our APIs don't return this data anyway.
I hope that gives a bit of light. if not I'm more than happy to elaborate more :)

Reviewer: Does this release make the lineage view in the file browser work for files created via a workflow file?

Reviewer 2: Yes, that should work.

Reviewer 3: only one way to find out!

Reviewer: I opened the PR deployment and tried to view lineage on my Academic Team demo, and it [didn't work](https://ci-renku-2914.dev.renku.ch/projects/laura.kinkead/academic-team-demo-sklearn/files/lineage/data/05_metrics/model_comparison.png). I also tried with the workflow demo I'd been testing with Mohammad before the metadata 2.0 release, and lineage [doesn't work there either](https://ci-renku-2914.dev.renku.ch/projects/laura.kinkead/workflow-file-demo/files/lineage/data/output/filtered.csv). It's not clear to me if there's a different way I should be testing it.

From the user's point of view, the lineage view not working for file-based workflows is a regression. I think we should be careful to only let that regression into a release if it's intentional- i.e. we're aware of the technical tradeoffs and decide to make a release anyway (and communicate the regression to users). 

@jachro Would you be willing to the bottom of this and get a sure answer?

Reviewer 2: Sorry @lokijuhy I've got a bit tendency to miss notifications from GH :)

So the problem with your project and lineage is caused by problems during `renku migrate` we call for each commit we process. I'm guessing it might be because of some incompatibilities between the renku 1.11.1 we use for triples generation and the one you used for the project. The error we received from the CLI looks like this:
```
File \"/usr/bin/renku\", line 8, in <module>; 
sys.exit(cli()); 
File \"/usr/lib/python3.10/site-packages/click/core.py\", line 1130, in __call__; 
return self.main(*args, **kwargs); 
File \"/usr/lib/python3.10/site-packages/renku/ui/cli/exception_handler.py\", line 133, in main; 
return super().main(*args, **kwargs); 
File \"/usr/lib/python3.10/site-packages/renku/ui/cli/exception_handler.py\", line 92, in main; 
return super().main(*args, **kwargs); 
File \"/usr/lib/python3.10/site-packages/click/core.py\", line 1055, in main; 
rv = self.invoke(ctx); 
File \"/usr/lib/python3.10/site-packages/click/core.py\", line 1657, in invoke; 
return _process_result(sub_ctx.command.invoke(sub_ctx)); 
File \"/usr/lib/python3.10/site-packages/click/core.py\", line 1404, in invoke; 
return ctx.invoke(self.callback, **ctx.params); 
File \"/usr/lib/python3.10/site-packages/click/core.py\", line 760, in invoke; 
return __callback(*args, **kwargs); 
File \"/usr/lib/python3.10/site-packages/renku/ui/cli/migrate.py\", line 98, in migrate; 
status = check_project().build().execute().output; 
File \"/usr/lib/python3.10/site-packages/renku/command/command_builder/command.py\", line 223, in execute; 
hook(self, context, *args, **kwargs); 
File \"/usr/lib/python3.10/site-packages/renku/command/command_builder/database.py\", line 84, in _injection_pre_hook; 
minimum_renku_version = Version(self.project.minimum_renku_version); 
File \"/usr/lib/python3.10/site-packages/renku/infrastructure/database.py\", line 435, in setstate; 
self._reader.set_ghost_state(object, data); 
File \"/usr/lib/python3.10/site-packages/renku/infrastructure/database.py\", line 973, in set_ghost_state; 
state = self._deserialize_helper(data, create=False); 
File \"/usr/lib/python3.10/site-packages/renku/infrastructure/database.py\", line 1066, in _deserialize_helper; 
data = self._deserialize_helper(data); 
File \"/usr/lib/python3.10/site-packages/renku/infrastructure/database.py\", line 1012, in _deserialize_helper; 
data[key] = self._deserialize_helper(value); 
File \"/usr/lib/python3.10/site-packages/renku/infrastructure/database.py\", line 1027, in _deserialize_helper; 
cls = self._get_class(object_type); 
File \"/usr/lib/python3.10/site-packages/renku/infrastructure/database.py\", line 955, in _get_class; 
cls = get_class(type_name); 
File \"/usr/lib/python3.10/site-packages/renku/infrastructure/database.py\", line 112, in get_class; 
return get_attribute(module, components[1:]); 
File \"/usr/lib/python3.10/site-packages/renku/infrastructure/database.py\", line 145, in get_attribute; 
object = getattr(object, component); 
AttributeError: module 'renku.domain_model.project' has no attribute 'ProjectTemplateMetadata';
```
@Panaetius maybe you can give us some help, please.

Reviewer: @jachro Ok! I just talked to Ralf, and he modified this CI deployment so I can make a test of WFF-lineage view. I will do that and report back the results!

.. _changelog:

0.23.0
------

Renku ``0.23.0`` introduces the Renku Workflow File, a friendlier way to encode and run your data analysis pipelines

> on Renku. This yaml-based file is easy to write and execution is as easy as `renku run workflow.yml`. Add a workflow

Reviewer 4: I guess the file is a yaml file and not a yaml-based file; the workflow is yaml-based, the approach is arguably yaml based.

How about:

` Workflows are based on an easy to use yaml file and execution is as easy as...`

Reviewer: ```suggestion
on Renku. You can write out your workflow in this easy-to-use YAML file, and execute it with `renku run workflow.yml`. Add a workflow
```

to your Renku project with our docs [here](https://renku.readthedocs.io/en/stable/topic-guides/workflows/workflow-file.html)!

User-Facing Changes
~~~~~~~~~~~~~~~~~~~

**🌟 New Features**

>  * 📝 **Renku CLI**: Add support for writing workflow files which can be executed directly, greatly improving the UX

Reviewer 4: Does the CLI add support for writing workflow files or just parsing and executing them?

Reviewer 5: both

Reviewer 3: I would leave the word "writing" out, if I understood @seanrmurphy 's comment correctly

Reviewer 4: I just cannot imagine what is meant here - write a file based on what? Does it pull it down from somewhere? where does it write it to? Is it some kind of user based interaction to generate the file?

Also, it's not completely clear to me whether the execution is part of the change (I guess not).

If writing a file is the essence of the change and execution already existed, I guess we can't really leave out writing.

Is this somehow what is meant?

`Add support for CLI-based interactive workflow file generation`

WDYT @Panaetius ?

Reviewer: @seanrmurphy Ralf already edited the line, it now reads `*Add support for workflow files which can be executed directly, greatly improving the UX around workflows`

   around workflows
   (`#3176 <https://github.com/SwissDataScienceCenter/renku-python/pull/3176>`_)

>  * **Knowledge Graph**: Partial support for Workflow File metadata.

Reviewer 4: I don't understand this - would either of these make sense? 

Add partial support for indexing workflows? 
Add partial support for processing workflow metadata?  (this is not really much better tbh in the sense that what this processing is is not clear at all...)

Reviewer 2: Yes, I see your point. Should we even drop this line altogether? From the user's point of view, it's crucial if things will work or not as it's not even possible that he/she will be able to see the difference between the so-called 'partial support' and full support if there is such. I mean, for the time being, the user wouldn't have access to the full-picture data even if it's there.

**✨ Improvements**

>  * ⚙️ **Renku CLI**: Made `toil` the default workflow backend instead of `cwl`

Reviewer: I think it's fine as it is
